{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scATAC-seq analysis"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set-up for running a scATAC-seq analysis with pycisTopic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import pycisTopic\n",
    "import pyranges as pr\n",
    "import requests\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull,'wb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_name = \"CharacterizationMcGinnis_Dataset6\"\n",
    "combinomics_dir = \"/cellar/users/aklie/data/igvf/topic_grn_links/combinomics\"\n",
    "results_dir = \"/cellar/users/aklie/projects/igvf/topic_grn_links/grn_inference/scenicplus/results\"\n",
    "dataset_name = \"CharacterizationMcGinnis_Dataset6\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a directory for to store the processed scATAC-seq data.\n",
    "if not os.path.exists(os.path.join(results_dir, dataset_name, 'scATAC')):\n",
    "    os.makedirs(os.path.join(results_dir, dataset_name, 'scATAC'))\n",
    "tmp_dir = '/cellar/users/aklie/tmp/'"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You will need to create a dictionary with values as paths to fragment files and keys as sample names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_1.atac.filter.fragments.hg38.tsv.gz',\n",
       " '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_2.atac.filter.fragments.hg38.tsv.gz',\n",
       " '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_3.atac.filter.fragments.hg38.tsv.gz',\n",
       " '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_4.atac.filter.fragments.hg38.tsv.gz']"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragment_files = os.path.join(combinomics_dir, dataset_name, \"*tsv.gz\")\n",
    "fragment_files = sorted(glob.glob(fragment_files))\n",
    "fragment_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Dataset6_sample_1': '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_1.atac.filter.fragments.hg38.tsv.gz',\n",
       " 'Dataset6_sample_2': '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_2.atac.filter.fragments.hg38.tsv.gz',\n",
       " 'Dataset6_sample_3': '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_3.atac.filter.fragments.hg38.tsv.gz',\n",
       " 'Dataset6_sample_4': '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_4.atac.filter.fragments.hg38.tsv.gz'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Load in the paths to your fragment files and turn them into a dictionary\n",
    "fragments_dict = dict(zip([os.path.basename(file).split(\".\")[0] for file in fragment_files], fragment_files))\n",
    "fragments_dict"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in or grab the metadata from any previous methods for annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_1.atac.qc.hg38.metadata.tsv',\n",
       " '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_2.atac.qc.hg38.metadata.tsv',\n",
       " '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_3.atac.qc.hg38.metadata.tsv',\n",
       " '/cellar/users/aklie/data/igvf/topic_grn_links/combinomics/CharacterizationMcGinnis_Dataset6/Dataset6_sample_4.atac.qc.hg38.metadata.tsv']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_metadata_glob = os.path.join(combinomics_dir, dataset_name, '*atac.qc.hg38.metadata.tsv')\n",
    "atac_metadata_files = sorted(glob.glob(atac_metadata_glob))\n",
    "atac_metadata_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.concat([pd.read_csv(f, sep='\\t', index_col=0) for f in atac_metadata_files])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>RNA_snn_res.0.25</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>PercentMito</th>\n",
       "      <th>lane</th>\n",
       "      <th>MULTI</th>\n",
       "      <th>RNA_snn_res.1</th>\n",
       "      <th>RNA_snn_res.0.5</th>\n",
       "      <th>sample</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAGGTAAC_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>5710</td>\n",
       "      <td>2647</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.066900</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar64</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>S5_2D</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACCTCGCT_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>7103</td>\n",
       "      <td>2723</td>\n",
       "      <td>13</td>\n",
       "      <td>21</td>\n",
       "      <td>0.059693</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>21</td>\n",
       "      <td>13</td>\n",
       "      <td>S4</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACTCGCTC_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>6402</td>\n",
       "      <td>2918</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.037176</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar64</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>S5_2D</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGCAATAA_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>1721</td>\n",
       "      <td>991</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.160372</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar65</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>S5_2D3D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGCGCTTG_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>2853</td>\n",
       "      <td>1535</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>0.049772</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar66</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>S6_2D</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTCCAAA_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>5090</td>\n",
       "      <td>2357</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.038310</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar66</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>S6_2D</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTCCTGC_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>3570</td>\n",
       "      <td>1924</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0.001681</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar64</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>S5_2D</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTTTGTC_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>1093</td>\n",
       "      <td>890</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.020128</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>S4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAAGCGC_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>5066</td>\n",
       "      <td>2621</td>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "      <td>0.042835</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar65</td>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>S5_2D3D</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTGGCCGA_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>2989</td>\n",
       "      <td>1863</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>0.055872</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar61</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>S1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72152 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1  SeuratProject        5710          2647   \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1  SeuratProject        7103          2723   \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1  SeuratProject        6402          2918   \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1  SeuratProject        1721           991   \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1  SeuratProject        2853          1535   \n",
       "...                                           ...         ...           ...   \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4  SeuratProject        5090          2357   \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4  SeuratProject        3570          1924   \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4  SeuratProject        1093           890   \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4  SeuratProject        5066          2621   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4  SeuratProject        2989          1863   \n",
       "\n",
       "                                    RNA_snn_res.0.25  seurat_clusters  \\\n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1                10                9   \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1                13               21   \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1                 3                2   \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1                 7                3   \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1                 5               15   \n",
       "...                                              ...              ...   \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4                 2                0   \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4                 6                6   \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4                 0                7   \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4                10               24   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4                 9                8   \n",
       "\n",
       "                                    PercentMito  lane  MULTI  RNA_snn_res.1  \\\n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1     0.066900     1  Bar64              9   \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1     0.059693     1  Bar63             21   \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1     0.037176     1  Bar64              2   \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1     0.160372     1  Bar65              3   \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1     0.049772     1  Bar66             15   \n",
       "...                                         ...   ...    ...            ...   \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4     0.038310     4  Bar66              0   \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4     0.001681     4  Bar64              6   \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4     0.020128     4  Bar63              7   \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4     0.042835     4  Bar65             24   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4     0.055872     4  Bar61              8   \n",
       "\n",
       "                                    RNA_snn_res.0.5   sample  temp  \n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1                8    S5_2D     9  \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1               13       S4    21  \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1                4    S5_2D     2  \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1                6  S5_2D3D     3  \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1                2    S6_2D    15  \n",
       "...                                             ...      ...   ...  \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4                1    S6_2D     0  \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4                5    S5_2D     6  \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4               10       S4     7  \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4                8  S5_2D3D    24  \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4                9       S1     8  \n",
       "\n",
       "[72152 rows x 12 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Chris McGinnis annotations\n",
    "metadata_dir = \"/cellar/users/aklie/data/igvf/topic_grn_links/metadata\"\n",
    "processed_metadata_file = os.path.join(metadata_dir, dataset_name, 'metadata_IGVF6.csv')\n",
    "processed_metadata = pd.read_csv(processed_metadata_file, index_col=0)\n",
    "index_split = processed_metadata.index.str.split(\"-\")\n",
    "processed_metadata.index = [x[0] + \"_Dataset6_sample_\" + x[1] for x in index_split]\n",
    "processed_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fragments_promoter</th>\n",
       "      <th>reads_tss</th>\n",
       "      <th>reads_promoter</th>\n",
       "      <th>tss_enrichment</th>\n",
       "      <th>reads_tss_total</th>\n",
       "      <th>total</th>\n",
       "      <th>duplicate</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>lowmapq</th>\n",
       "      <th>unique</th>\n",
       "      <th>...</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>RNA_snn_res.0.25</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>PercentMito</th>\n",
       "      <th>lane</th>\n",
       "      <th>MULTI</th>\n",
       "      <th>RNA_snn_res.1</th>\n",
       "      <th>RNA_snn_res.0.5</th>\n",
       "      <th>sample</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAACGCGA_Dataset6_sample_4</th>\n",
       "      <td>17088</td>\n",
       "      <td>3423</td>\n",
       "      <td>33940</td>\n",
       "      <td>14.033577</td>\n",
       "      <td>3423.0</td>\n",
       "      <td>10210</td>\n",
       "      <td>2872</td>\n",
       "      <td>754</td>\n",
       "      <td>0</td>\n",
       "      <td>6584</td>\n",
       "      <td>...</td>\n",
       "      <td>457</td>\n",
       "      <td>7</td>\n",
       "      <td>23</td>\n",
       "      <td>0.187793</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>23</td>\n",
       "      <td>12</td>\n",
       "      <td>S4</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACCTAAT_Dataset6_sample_3</th>\n",
       "      <td>28363</td>\n",
       "      <td>6324</td>\n",
       "      <td>56294</td>\n",
       "      <td>14.527578</td>\n",
       "      <td>6324.0</td>\n",
       "      <td>14754</td>\n",
       "      <td>4991</td>\n",
       "      <td>830</td>\n",
       "      <td>0</td>\n",
       "      <td>8933</td>\n",
       "      <td>...</td>\n",
       "      <td>2503</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.082565</td>\n",
       "      <td>3</td>\n",
       "      <td>Bar66</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>S6_2D</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACGTGCT_Dataset6_sample_2</th>\n",
       "      <td>1989</td>\n",
       "      <td>370</td>\n",
       "      <td>3937</td>\n",
       "      <td>11.817311</td>\n",
       "      <td>370.0</td>\n",
       "      <td>1280</td>\n",
       "      <td>341</td>\n",
       "      <td>121</td>\n",
       "      <td>0</td>\n",
       "      <td>818</td>\n",
       "      <td>...</td>\n",
       "      <td>4350</td>\n",
       "      <td>11</td>\n",
       "      <td>25</td>\n",
       "      <td>0.083430</td>\n",
       "      <td>2</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>25</td>\n",
       "      <td>16</td>\n",
       "      <td>S4</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAACTGGCT_Dataset6_sample_2</th>\n",
       "      <td>20899</td>\n",
       "      <td>4449</td>\n",
       "      <td>41451</td>\n",
       "      <td>14.141093</td>\n",
       "      <td>4449.0</td>\n",
       "      <td>10961</td>\n",
       "      <td>3392</td>\n",
       "      <td>581</td>\n",
       "      <td>0</td>\n",
       "      <td>6988</td>\n",
       "      <td>...</td>\n",
       "      <td>2168</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.061338</td>\n",
       "      <td>2</td>\n",
       "      <td>Bar66</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>S6_2D</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAGCCAGA_Dataset6_sample_4</th>\n",
       "      <td>42589</td>\n",
       "      <td>9141</td>\n",
       "      <td>84417</td>\n",
       "      <td>15.021569</td>\n",
       "      <td>9141.0</td>\n",
       "      <td>21966</td>\n",
       "      <td>6166</td>\n",
       "      <td>1153</td>\n",
       "      <td>0</td>\n",
       "      <td>14647</td>\n",
       "      <td>...</td>\n",
       "      <td>3478</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>0.054582</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar66</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>S6_2D</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTGGCCGA_Dataset6_sample_4</th>\n",
       "      <td>68800</td>\n",
       "      <td>13642</td>\n",
       "      <td>136343</td>\n",
       "      <td>12.724381</td>\n",
       "      <td>13642.0</td>\n",
       "      <td>37133</td>\n",
       "      <td>9709</td>\n",
       "      <td>3032</td>\n",
       "      <td>0</td>\n",
       "      <td>24392</td>\n",
       "      <td>...</td>\n",
       "      <td>1863</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>0.055872</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar61</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>S1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTAACGG_Dataset6_sample_1</th>\n",
       "      <td>90339</td>\n",
       "      <td>18836</td>\n",
       "      <td>179149</td>\n",
       "      <td>15.021752</td>\n",
       "      <td>18836.0</td>\n",
       "      <td>63967</td>\n",
       "      <td>24015</td>\n",
       "      <td>3805</td>\n",
       "      <td>0</td>\n",
       "      <td>36147</td>\n",
       "      <td>...</td>\n",
       "      <td>1505</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>0.006016</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>20</td>\n",
       "      <td>7</td>\n",
       "      <td>S4</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTATTCG_Dataset6_sample_3</th>\n",
       "      <td>39529</td>\n",
       "      <td>9556</td>\n",
       "      <td>78510</td>\n",
       "      <td>21.901357</td>\n",
       "      <td>9556.0</td>\n",
       "      <td>21712</td>\n",
       "      <td>6899</td>\n",
       "      <td>1339</td>\n",
       "      <td>0</td>\n",
       "      <td>13474</td>\n",
       "      <td>...</td>\n",
       "      <td>1681</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>0.047019</td>\n",
       "      <td>3</td>\n",
       "      <td>Bar64</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>S5_2D</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGCAGA_Dataset6_sample_2</th>\n",
       "      <td>27188</td>\n",
       "      <td>5644</td>\n",
       "      <td>53924</td>\n",
       "      <td>16.484126</td>\n",
       "      <td>5644.0</td>\n",
       "      <td>14667</td>\n",
       "      <td>4342</td>\n",
       "      <td>905</td>\n",
       "      <td>0</td>\n",
       "      <td>9420</td>\n",
       "      <td>...</td>\n",
       "      <td>1046</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.034535</td>\n",
       "      <td>2</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>S4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGCTT_Dataset6_sample_3</th>\n",
       "      <td>66185</td>\n",
       "      <td>13965</td>\n",
       "      <td>131104</td>\n",
       "      <td>15.278158</td>\n",
       "      <td>13965.0</td>\n",
       "      <td>40163</td>\n",
       "      <td>12525</td>\n",
       "      <td>2308</td>\n",
       "      <td>0</td>\n",
       "      <td>25330</td>\n",
       "      <td>...</td>\n",
       "      <td>1253</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>0.157895</td>\n",
       "      <td>3</td>\n",
       "      <td>Bar67</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>S6_2D3D</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72104 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    fragments_promoter  reads_tss  \\\n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4               17088       3423   \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3               28363       6324   \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2                1989        370   \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2               20899       4449   \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4               42589       9141   \n",
       "...                                                ...        ...   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4               68800      13642   \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1               90339      18836   \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3               39529       9556   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2               27188       5644   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3               66185      13965   \n",
       "\n",
       "                                    reads_promoter  tss_enrichment  \\\n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4           33940       14.033577   \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3           56294       14.527578   \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2            3937       11.817311   \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2           41451       14.141093   \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4           84417       15.021569   \n",
       "...                                            ...             ...   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4          136343       12.724381   \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1          179149       15.021752   \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3           78510       21.901357   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2           53924       16.484126   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3          131104       15.278158   \n",
       "\n",
       "                                    reads_tss_total  total  duplicate  \\\n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4           3423.0  10210       2872   \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3           6324.0  14754       4991   \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2            370.0   1280        341   \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2           4449.0  10961       3392   \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4           9141.0  21966       6166   \n",
       "...                                             ...    ...        ...   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4          13642.0  37133       9709   \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1          18836.0  63967      24015   \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3           9556.0  21712       6899   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2           5644.0  14667       4342   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3          13965.0  40163      12525   \n",
       "\n",
       "                                    unmapped  lowmapq  unique  ...  \\\n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4       754        0    6584  ...   \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3       830        0    8933  ...   \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2       121        0     818  ...   \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2       581        0    6988  ...   \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4      1153        0   14647  ...   \n",
       "...                                      ...      ...     ...  ...   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4      3032        0   24392  ...   \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1      3805        0   36147  ...   \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3      1339        0   13474  ...   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2       905        0    9420  ...   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3      2308        0   25330  ...   \n",
       "\n",
       "                                    nFeature_RNA  RNA_snn_res.0.25  \\\n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4           457                 7   \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3          2503                 1   \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2          4350                11   \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2          2168                 1   \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4          3478                 1   \n",
       "...                                          ...               ...   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4          1863                 9   \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1          1505                 8   \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3          1681                 4   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2          1046                 0   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3          1253                 5   \n",
       "\n",
       "                                    seurat_clusters  PercentMito  lane  MULTI  \\\n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4               23     0.187793     4  Bar63   \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3               12     0.082565     3  Bar66   \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2               25     0.083430     2  Bar63   \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2                1     0.061338     2  Bar66   \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4               12     0.054582     4  Bar66   \n",
       "...                                             ...          ...   ...    ...   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4                8     0.055872     4  Bar61   \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1               20     0.006016     1  Bar63   \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3                4     0.047019     3  Bar64   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2                7     0.034535     2  Bar63   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3               15     0.157895     3  Bar67   \n",
       "\n",
       "                                    RNA_snn_res.1 RNA_snn_res.0.5   sample  \\\n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4             23              12       S4   \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3             12               0    S6_2D   \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2             25              16       S4   \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2              1               0    S6_2D   \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4             12               0    S6_2D   \n",
       "...                                           ...             ...      ...   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4              8               9       S1   \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1             20               7       S4   \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3              4               3    S5_2D   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2              7              10       S4   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3             15               2  S6_2D3D   \n",
       "\n",
       "                                    temp  \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_4    23  \n",
       "AAACAGCCAACCTAAT_Dataset6_sample_3    12  \n",
       "AAACAGCCAACGTGCT_Dataset6_sample_2    25  \n",
       "AAACAGCCAACTGGCT_Dataset6_sample_2     1  \n",
       "AAACAGCCAAGCCAGA_Dataset6_sample_4    12  \n",
       "...                                  ...  \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4     8  \n",
       "TTTGTTGGTTTAACGG_Dataset6_sample_1    20  \n",
       "TTTGTTGGTTTATTCG_Dataset6_sample_3     4  \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_2     7  \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_3    15  \n",
       "\n",
       "[72104 rows x 29 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.merge(processed_metadata, left_index=True, right_index=True, how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>RNA_snn_res.0.25</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>PercentMito</th>\n",
       "      <th>lane</th>\n",
       "      <th>MULTI</th>\n",
       "      <th>RNA_snn_res.1</th>\n",
       "      <th>RNA_snn_res.0.5</th>\n",
       "      <th>sample</th>\n",
       "      <th>temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAGGTAAC_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>5710</td>\n",
       "      <td>2647</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.066900</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar64</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>S5_2D</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACCTCGCT_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>7103</td>\n",
       "      <td>2723</td>\n",
       "      <td>13</td>\n",
       "      <td>21</td>\n",
       "      <td>0.059693</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>21</td>\n",
       "      <td>13</td>\n",
       "      <td>S4</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCACTCGCTC_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>6402</td>\n",
       "      <td>2918</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>0.037176</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar64</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>S5_2D</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGCAATAA_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>1721</td>\n",
       "      <td>991</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>0.160372</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar65</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>S5_2D3D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAGCGCTTG_Dataset6_sample_1</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>2853</td>\n",
       "      <td>1535</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>0.049772</td>\n",
       "      <td>1</td>\n",
       "      <td>Bar66</td>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>S6_2D</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTCCAAA_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>5090</td>\n",
       "      <td>2357</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.038310</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar66</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>S6_2D</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTCCTGC_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>3570</td>\n",
       "      <td>1924</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>0.001681</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar64</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>S5_2D</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTTTGTC_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>1093</td>\n",
       "      <td>890</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0.020128</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar63</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "      <td>S4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTAAGCGC_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>5066</td>\n",
       "      <td>2621</td>\n",
       "      <td>10</td>\n",
       "      <td>24</td>\n",
       "      <td>0.042835</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar65</td>\n",
       "      <td>24</td>\n",
       "      <td>8</td>\n",
       "      <td>S5_2D3D</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTGGCCGA_Dataset6_sample_4</th>\n",
       "      <td>SeuratProject</td>\n",
       "      <td>2989</td>\n",
       "      <td>1863</td>\n",
       "      <td>9</td>\n",
       "      <td>8</td>\n",
       "      <td>0.055872</td>\n",
       "      <td>4</td>\n",
       "      <td>Bar61</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>S1</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72152 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1  SeuratProject        5710          2647   \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1  SeuratProject        7103          2723   \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1  SeuratProject        6402          2918   \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1  SeuratProject        1721           991   \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1  SeuratProject        2853          1535   \n",
       "...                                           ...         ...           ...   \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4  SeuratProject        5090          2357   \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4  SeuratProject        3570          1924   \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4  SeuratProject        1093           890   \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4  SeuratProject        5066          2621   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4  SeuratProject        2989          1863   \n",
       "\n",
       "                                    RNA_snn_res.0.25  seurat_clusters  \\\n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1                10                9   \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1                13               21   \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1                 3                2   \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1                 7                3   \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1                 5               15   \n",
       "...                                              ...              ...   \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4                 2                0   \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4                 6                6   \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4                 0                7   \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4                10               24   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4                 9                8   \n",
       "\n",
       "                                    PercentMito  lane  MULTI  RNA_snn_res.1  \\\n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1     0.066900     1  Bar64              9   \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1     0.059693     1  Bar63             21   \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1     0.037176     1  Bar64              2   \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1     0.160372     1  Bar65              3   \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1     0.049772     1  Bar66             15   \n",
       "...                                         ...   ...    ...            ...   \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4     0.038310     4  Bar66              0   \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4     0.001681     4  Bar64              6   \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4     0.020128     4  Bar63              7   \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4     0.042835     4  Bar65             24   \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4     0.055872     4  Bar61              8   \n",
       "\n",
       "                                    RNA_snn_res.0.5   sample  temp  \n",
       "AAACAGCCAAGGTAAC_Dataset6_sample_1                8    S5_2D     9  \n",
       "AAACAGCCACCTCGCT_Dataset6_sample_1               13       S4    21  \n",
       "AAACAGCCACTCGCTC_Dataset6_sample_1                4    S5_2D     2  \n",
       "AAACAGCCAGCAATAA_Dataset6_sample_1                6  S5_2D3D     3  \n",
       "AAACAGCCAGCGCTTG_Dataset6_sample_1                2    S6_2D    15  \n",
       "...                                             ...      ...   ...  \n",
       "TTTGTTGGTGTCCAAA_Dataset6_sample_4                1    S6_2D     0  \n",
       "TTTGTTGGTGTCCTGC_Dataset6_sample_4                5    S5_2D     6  \n",
       "TTTGTTGGTGTTTGTC_Dataset6_sample_4               10       S4     7  \n",
       "TTTGTTGGTTAAGCGC_Dataset6_sample_4                8  S5_2D3D    24  \n",
       "TTTGTTGGTTGGCCGA_Dataset6_sample_4                9       S1     8  \n",
       "\n",
       "[72152 rows x 12 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed_metadata"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Map the RNA barcodes from the anndata to the ATAC barcodes present in the fragment file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_fragments = pd.read_csv(fragments_dict[\"Dataset6_sample_4\"], sep=\"\\t\", header=None, nrows=1000)\n",
    "metadata.index.isin(test_fragments[3]).sum()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also need chromosome sizes in order to export to pseudobulk and to call peaks. Read these in. Make sure you choose the correct species for your data!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stream directly into memory using pandas\n",
    "target_url = 'http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes'\n",
    "chromsizes = pd.read_csv(target_url, sep='\\t', header=None)\n",
    "chromsizes.columns = ['Chromosome', 'End']\n",
    "chromsizes['Start'] = [0]*chromsizes.shape[0]\n",
    "chromsizes = chromsizes.loc[:,['Chromosome', 'Start', 'End']]\n",
    "chromsizes = pr.PyRanges(chromsizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_GL383518v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>182439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_GL383519v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>110268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_GL383520v2_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>366580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_KI270706v1_random</td>\n",
       "      <td>0</td>\n",
       "      <td>175055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>chrX_KI270880v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>284869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>chrX_KI270881v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>144206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>chrX_KI270913v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>274009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>chrY</td>\n",
       "      <td>0</td>\n",
       "      <td>57227415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>0</td>\n",
       "      <td>37240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>455 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+------------------------+-----------+-----------+\n",
       "| Chromosome             | Start     | End       |\n",
       "| (category)             | (int32)   | (int32)   |\n",
       "|------------------------+-----------+-----------|\n",
       "| chr1                   | 0         | 248956422 |\n",
       "| chr1_GL383518v1_alt    | 0         | 182439    |\n",
       "| chr1_GL383519v1_alt    | 0         | 110268    |\n",
       "| chr1_GL383520v2_alt    | 0         | 366580    |\n",
       "| ...                    | ...       | ...       |\n",
       "| chrX_KI270881v1_alt    | 0         | 144206    |\n",
       "| chrX_KI270913v1_alt    | 0         | 274009    |\n",
       "| chrY                   | 0         | 57227415  |\n",
       "| chrY_KI270740v1_random | 0         | 37240     |\n",
       "+------------------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 455 rows and 3 columns from 455 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromsizes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next big step is to generate bed and bw files using all the fragments within each cell type. pycisTopic has a decent function for this that's a little finicky. A couple things to make sure of:\n",
    "- You have a \"barcode\" column with barcodes that match the barcodes in your fragment files\n",
    "- You have a \"sample_id\" column that maps cells to the samples contained in your fragment file dictionary\n",
    "- Your chromosomes in the chromsizes object you created match the chromsomes in your fragment files (make sure you have the right species)\n",
    "- You have enough total memory to handle ray's parallelization (if you use it)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fragments_promoter</th>\n",
       "      <th>reads_tss</th>\n",
       "      <th>reads_promoter</th>\n",
       "      <th>tss_enrichment</th>\n",
       "      <th>reads_tss_total</th>\n",
       "      <th>total</th>\n",
       "      <th>duplicate</th>\n",
       "      <th>unmapped</th>\n",
       "      <th>lowmapq</th>\n",
       "      <th>unique</th>\n",
       "      <th>pct_dup</th>\n",
       "      <th>pct_unmapped</th>\n",
       "      <th>reads_peaks</th>\n",
       "      <th>fragment_peaks</th>\n",
       "      <th>pct_reads_promoter</th>\n",
       "      <th>pct_reads_peaks</th>\n",
       "      <th>pct_mito_reads</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAACCTTG_Dataset6_sample_1</th>\n",
       "      <td>939</td>\n",
       "      <td>132</td>\n",
       "      <td>1849</td>\n",
       "      <td>3.308685</td>\n",
       "      <td>132.0</td>\n",
       "      <td>1366</td>\n",
       "      <td>440</td>\n",
       "      <td>218</td>\n",
       "      <td>0</td>\n",
       "      <td>708</td>\n",
       "      <td>0.383275</td>\n",
       "      <td>0.159590</td>\n",
       "      <td>290</td>\n",
       "      <td>187</td>\n",
       "      <td>135.359</td>\n",
       "      <td>0</td>\n",
       "      <td>29.3977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAACGCGA_Dataset6_sample_1</th>\n",
       "      <td>28</td>\n",
       "      <td>4</td>\n",
       "      <td>56</td>\n",
       "      <td>0.396040</td>\n",
       "      <td>4.0</td>\n",
       "      <td>31</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.096774</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>180.645</td>\n",
       "      <td>0</td>\n",
       "      <td>40.3846</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAACGGGC_Dataset6_sample_1</th>\n",
       "      <td>105</td>\n",
       "      <td>20</td>\n",
       "      <td>210</td>\n",
       "      <td>1.980198</td>\n",
       "      <td>20.0</td>\n",
       "      <td>60</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>51</td>\n",
       "      <td>0.072727</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>40</td>\n",
       "      <td>23</td>\n",
       "      <td>350.000</td>\n",
       "      <td>0</td>\n",
       "      <td>53.3980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAAGCGCA_Dataset6_sample_1</th>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>57</td>\n",
       "      <td>0.099010</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0.421053</td>\n",
       "      <td>0.050000</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>285.000</td>\n",
       "      <td>0</td>\n",
       "      <td>10.6145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCCAAAGCTAA_Dataset6_sample_1</th>\n",
       "      <td>43</td>\n",
       "      <td>4</td>\n",
       "      <td>83</td>\n",
       "      <td>0.396040</td>\n",
       "      <td>4.0</td>\n",
       "      <td>41</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>0.382353</td>\n",
       "      <td>0.170732</td>\n",
       "      <td>28</td>\n",
       "      <td>18</td>\n",
       "      <td>202.439</td>\n",
       "      <td>0</td>\n",
       "      <td>30.8690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGCAGA_Dataset6_sample_4</th>\n",
       "      <td>67</td>\n",
       "      <td>13</td>\n",
       "      <td>132</td>\n",
       "      <td>1.287129</td>\n",
       "      <td>13.0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.217391</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>573.913</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGCTT_Dataset6_sample_4</th>\n",
       "      <td>99</td>\n",
       "      <td>4</td>\n",
       "      <td>196</td>\n",
       "      <td>0.396040</td>\n",
       "      <td>4.0</td>\n",
       "      <td>59</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>0.021739</td>\n",
       "      <td>0.220339</td>\n",
       "      <td>33</td>\n",
       "      <td>23</td>\n",
       "      <td>332.203</td>\n",
       "      <td>0</td>\n",
       "      <td>91.0198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGGCG_Dataset6_sample_4</th>\n",
       "      <td>32</td>\n",
       "      <td>17</td>\n",
       "      <td>63</td>\n",
       "      <td>1.683168</td>\n",
       "      <td>17.0</td>\n",
       "      <td>25</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>252.000</td>\n",
       "      <td>0</td>\n",
       "      <td>57.1429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGGGTA_Dataset6_sample_4</th>\n",
       "      <td>19</td>\n",
       "      <td>10</td>\n",
       "      <td>38</td>\n",
       "      <td>0.990099</td>\n",
       "      <td>10.0</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.062500</td>\n",
       "      <td>13</td>\n",
       "      <td>7</td>\n",
       "      <td>237.500</td>\n",
       "      <td>0</td>\n",
       "      <td>31.9149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTTTGTCTA_Dataset6_sample_4</th>\n",
       "      <td>401</td>\n",
       "      <td>64</td>\n",
       "      <td>797</td>\n",
       "      <td>6.336634</td>\n",
       "      <td>64.0</td>\n",
       "      <td>399</td>\n",
       "      <td>97</td>\n",
       "      <td>44</td>\n",
       "      <td>0</td>\n",
       "      <td>258</td>\n",
       "      <td>0.273239</td>\n",
       "      <td>0.110276</td>\n",
       "      <td>98</td>\n",
       "      <td>64</td>\n",
       "      <td>199.749</td>\n",
       "      <td>0</td>\n",
       "      <td>28.7540</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1239392 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    fragments_promoter  reads_tss  \\\n",
       "barcode                                                             \n",
       "AAACAGCCAAACCTTG_Dataset6_sample_1                 939        132   \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_1                  28          4   \n",
       "AAACAGCCAAACGGGC_Dataset6_sample_1                 105         20   \n",
       "AAACAGCCAAAGCGCA_Dataset6_sample_1                  29          1   \n",
       "AAACAGCCAAAGCTAA_Dataset6_sample_1                  43          4   \n",
       "...                                                ...        ...   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_4                  67         13   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_4                  99          4   \n",
       "TTTGTTGGTTTGGGCG_Dataset6_sample_4                  32         17   \n",
       "TTTGTTGGTTTGGGTA_Dataset6_sample_4                  19         10   \n",
       "TTTGTTGGTTTGTCTA_Dataset6_sample_4                 401         64   \n",
       "\n",
       "                                    reads_promoter  tss_enrichment  \\\n",
       "barcode                                                              \n",
       "AAACAGCCAAACCTTG_Dataset6_sample_1            1849        3.308685   \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_1              56        0.396040   \n",
       "AAACAGCCAAACGGGC_Dataset6_sample_1             210        1.980198   \n",
       "AAACAGCCAAAGCGCA_Dataset6_sample_1              57        0.099010   \n",
       "AAACAGCCAAAGCTAA_Dataset6_sample_1              83        0.396040   \n",
       "...                                            ...             ...   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_4             132        1.287129   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_4             196        0.396040   \n",
       "TTTGTTGGTTTGGGCG_Dataset6_sample_4              63        1.683168   \n",
       "TTTGTTGGTTTGGGTA_Dataset6_sample_4              38        0.990099   \n",
       "TTTGTTGGTTTGTCTA_Dataset6_sample_4             797        6.336634   \n",
       "\n",
       "                                    reads_tss_total  total  duplicate  \\\n",
       "barcode                                                                 \n",
       "AAACAGCCAAACCTTG_Dataset6_sample_1            132.0   1366        440   \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_1              4.0     31          4   \n",
       "AAACAGCCAAACGGGC_Dataset6_sample_1             20.0     60          4   \n",
       "AAACAGCCAAAGCGCA_Dataset6_sample_1              1.0     20          8   \n",
       "AAACAGCCAAAGCTAA_Dataset6_sample_1              4.0     41         13   \n",
       "...                                             ...    ...        ...   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_4             13.0     23          0   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_4              4.0     59          1   \n",
       "TTTGTTGGTTTGGGCG_Dataset6_sample_4             17.0     25          3   \n",
       "TTTGTTGGTTTGGGTA_Dataset6_sample_4             10.0     16          2   \n",
       "TTTGTTGGTTTGTCTA_Dataset6_sample_4             64.0    399         97   \n",
       "\n",
       "                                    unmapped  lowmapq  unique   pct_dup  \\\n",
       "barcode                                                                   \n",
       "AAACAGCCAAACCTTG_Dataset6_sample_1       218        0     708  0.383275   \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_1         3        0      24  0.142857   \n",
       "AAACAGCCAAACGGGC_Dataset6_sample_1         5        0      51  0.072727   \n",
       "AAACAGCCAAAGCGCA_Dataset6_sample_1         1        0      11  0.421053   \n",
       "AAACAGCCAAAGCTAA_Dataset6_sample_1         7        0      21  0.382353   \n",
       "...                                      ...      ...     ...       ...   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_4         5        0      18  0.000000   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_4        13        0      45  0.021739   \n",
       "TTTGTTGGTTTGGGCG_Dataset6_sample_4         5        0      17  0.150000   \n",
       "TTTGTTGGTTTGGGTA_Dataset6_sample_4         1        0      13  0.133333   \n",
       "TTTGTTGGTTTGTCTA_Dataset6_sample_4        44        0     258  0.273239   \n",
       "\n",
       "                                    pct_unmapped  reads_peaks  fragment_peaks  \\\n",
       "barcode                                                                         \n",
       "AAACAGCCAAACCTTG_Dataset6_sample_1      0.159590          290             187   \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_1      0.096774           16              11   \n",
       "AAACAGCCAAACGGGC_Dataset6_sample_1      0.083333           40              23   \n",
       "AAACAGCCAAAGCGCA_Dataset6_sample_1      0.050000            2               1   \n",
       "AAACAGCCAAAGCTAA_Dataset6_sample_1      0.170732           28              18   \n",
       "...                                          ...          ...             ...   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_4      0.217391           10               6   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_4      0.220339           33              23   \n",
       "TTTGTTGGTTTGGGCG_Dataset6_sample_4      0.200000            8               6   \n",
       "TTTGTTGGTTTGGGTA_Dataset6_sample_4      0.062500           13               7   \n",
       "TTTGTTGGTTTGTCTA_Dataset6_sample_4      0.110276           98              64   \n",
       "\n",
       "                                    pct_reads_promoter  pct_reads_peaks  \\\n",
       "barcode                                                                   \n",
       "AAACAGCCAAACCTTG_Dataset6_sample_1             135.359                0   \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_1             180.645                0   \n",
       "AAACAGCCAAACGGGC_Dataset6_sample_1             350.000                0   \n",
       "AAACAGCCAAAGCGCA_Dataset6_sample_1             285.000                0   \n",
       "AAACAGCCAAAGCTAA_Dataset6_sample_1             202.439                0   \n",
       "...                                                ...              ...   \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_4             573.913                0   \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_4             332.203                0   \n",
       "TTTGTTGGTTTGGGCG_Dataset6_sample_4             252.000                0   \n",
       "TTTGTTGGTTTGGGTA_Dataset6_sample_4             237.500                0   \n",
       "TTTGTTGGTTTGTCTA_Dataset6_sample_4             199.749                0   \n",
       "\n",
       "                                    pct_mito_reads  \n",
       "barcode                                             \n",
       "AAACAGCCAAACCTTG_Dataset6_sample_1         29.3977  \n",
       "AAACAGCCAAACGCGA_Dataset6_sample_1         40.3846  \n",
       "AAACAGCCAAACGGGC_Dataset6_sample_1         53.3980  \n",
       "AAACAGCCAAAGCGCA_Dataset6_sample_1         10.6145  \n",
       "AAACAGCCAAAGCTAA_Dataset6_sample_1         30.8690  \n",
       "...                                            ...  \n",
       "TTTGTTGGTTTGCAGA_Dataset6_sample_4        100.0000  \n",
       "TTTGTTGGTTTGGCTT_Dataset6_sample_4         91.0198  \n",
       "TTTGTTGGTTTGGGCG_Dataset6_sample_4         57.1429  \n",
       "TTTGTTGGTTTGGGTA_Dataset6_sample_4         31.9149  \n",
       "TTTGTTGGTTTGTCTA_Dataset6_sample_4         28.7540  \n",
       "\n",
       "[1239392 rows x 17 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-01-14 22:54:59,585 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF187VMN/encode_scatac_dcc_2/results/ENCSR525WPH-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 22:56:44,138 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF035SPT/encode_scatac_dcc_2/results/ENCSR713FPX-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 22:58:25,383 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF622EUO/encode_scatac_dcc_2/results/ENCSR858YSB-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 23:00:24,434 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF119IVK/encode_scatac_dcc_2/results/ENCSR400PXQ-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 23:01:55,753 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF683IBE/encode_scatac_dcc_2/results/ENCSR668VPD-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 23:03:56,191 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF042ZJI/encode_scatac_dcc_2/results/ENCSR736ABS-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 23:05:45,716 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF176LJV/encode_scatac_dcc_2/results/ENCSR834NRP-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 23:07:43,911 cisTopic     INFO     Reading fragments from /cellar/users/aklie/data/igvf/topic_grn_links/mouse_adrenal/encode/fragments/ENCFF101BLM/encode_scatac_dcc_2/results/ENCSR024LKR-1/fragments/fragments.tsv.gz\n",
      "2023-01-14 23:09:27,214 cisTopic     INFO     Creating pseudobulk for Adipocytes\n",
      "2023-01-14 23:13:44,808 cisTopic     INFO     Adipocytes done!\n",
      "2023-01-14 23:13:44,992 cisTopic     INFO     Creating pseudobulk for Capsule\n",
      "2023-01-14 23:13:57,686 cisTopic     INFO     Capsule done!\n",
      "2023-01-14 23:13:57,695 cisTopic     INFO     Creating pseudobulk for Cortex\n",
      "2023-01-15 00:59:03,988 cisTopic     INFO     Cortex done!\n",
      "2023-01-15 00:59:08,161 cisTopic     INFO     Creating pseudobulk for Endothelial\n",
      "2023-01-15 01:12:19,987 cisTopic     INFO     Endothelial done!\n",
      "2023-01-15 01:12:20,496 cisTopic     INFO     Creating pseudobulk for Fibroblast\n",
      "2023-01-15 01:14:54,515 cisTopic     INFO     Fibroblast done!\n",
      "2023-01-15 01:14:54,575 cisTopic     INFO     Creating pseudobulk for Hepatocyte\n",
      "2023-01-15 01:15:01,065 cisTopic     INFO     Hepatocyte done!\n",
      "2023-01-15 01:15:01,067 cisTopic     INFO     Creating pseudobulk for Macrophages\n",
      "2023-01-15 01:16:52,776 cisTopic     INFO     Macrophages done!\n",
      "2023-01-15 01:16:52,819 cisTopic     INFO     Creating pseudobulk for Medulla\n",
      "2023-01-15 01:27:23,012 cisTopic     INFO     Medulla done!\n",
      "2023-01-15 01:27:23,200 cisTopic     INFO     Creating pseudobulk for Skeletal_muscle\n",
      "2023-01-15 01:27:27,106 cisTopic     INFO     Skeletal_muscle done!\n",
      "2023-01-15 01:27:27,107 cisTopic     INFO     Creating pseudobulk for Smooth_muscle\n",
      "2023-01-15 01:27:40,009 cisTopic     INFO     Smooth_muscle done!\n",
      "2023-01-15 01:27:40,015 cisTopic     INFO     Creating pseudobulk for Sox10_\n",
      "2023-01-15 01:28:20,816 cisTopic     INFO     Sox10_ done!\n",
      "2023-01-15 01:28:20,827 cisTopic     INFO     Creating pseudobulk for Stromal\n",
      "2023-01-15 01:32:55,037 cisTopic     INFO     Stromal done!\n"
     ]
    }
   ],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import export_pseudobulk\n",
    "\n",
    "# Run without ray since these are big files\n",
    "bw_paths, bed_paths = export_pseudobulk(\n",
    "    input_data = cell_data,\n",
    "    variable = 'celltype', # variable by which to generate pseubulk profiles, in this case we want pseudobulks per celltype\n",
    "    sample_id_col = 'sample_id',\n",
    "    chromsizes = chromsizes,\n",
    "    bed_path = os.path.join(work_dir, 'scATAC/consensus_peak_calling/pseudobulk_bed_files/'),  # specify where pseudobulk_bed_files should be stored\n",
    "    bigwig_path = os.path.join(work_dir, 'scATAC/consensus_peak_calling/pseudobulk_bw_files/'), # specify where pseudobulk_bw_files should be stored\n",
    "    path_to_fragments = fragments_dict, # location of fragment fiels\n",
    "    n_cpu = 1, # specify the number of cores to use, we use ray for multi processing\n",
    "    normalize_bigwig = True,\n",
    "    remove_duplicates = True,\n",
    "    _temp_dir = os.path.join(tmp_dir, 'ray_spill'),\n",
    "    split_pattern = '-'\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dump the paths to the bed and bw files for peak calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "# Export the paths\n",
    "pickle.dump(\n",
    "    bed_paths,\n",
    "    open(os.path.join(work_dir, 'scATAC/consensus_peak_calling/pseudobulk_bed_files/bed_paths.pkl'), 'wb')\n",
    ")\n",
    "pickle.dump(\n",
    "    bw_paths,\n",
    "    open(os.path.join(work_dir, 'scATAC/consensus_peak_calling/pseudobulk_bw_files/bw_paths.pkl'), 'wb')\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now run cell type specific peak calling using MACS2. pycisTopic has another useful function for this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-01-15 07:53:26,502\tINFO worker.py:1519 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044428)\u001b[0m 2023-01-15 07:53:39,567 cisTopic     INFO     Calling peaks for Stromal with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Stromal.bed.gz --name Stromal  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044425)\u001b[0m 2023-01-15 07:53:39,594 cisTopic     INFO     Calling peaks for Skeletal_muscle with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Skeletal_muscle.bed.gz --name Skeletal_muscle  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044422)\u001b[0m 2023-01-15 07:53:39,605 cisTopic     INFO     Calling peaks for Capsule with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Capsule.bed.gz --name Capsule  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044427)\u001b[0m 2023-01-15 07:53:39,579 cisTopic     INFO     Calling peaks for Medulla with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Medulla.bed.gz --name Medulla  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044431)\u001b[0m 2023-01-15 07:53:39,584 cisTopic     INFO     Calling peaks for Smooth_muscle with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Smooth_muscle.bed.gz --name Smooth_muscle  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044421)\u001b[0m 2023-01-15 07:53:39,669 cisTopic     INFO     Calling peaks for Cortex with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Cortex.bed.gz --name Cortex  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044430)\u001b[0m 2023-01-15 07:53:39,663 cisTopic     INFO     Calling peaks for Fibroblast with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Fibroblast.bed.gz --name Fibroblast  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044423)\u001b[0m 2023-01-15 07:53:39,654 cisTopic     INFO     Calling peaks for Sox10_ with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Sox10_.bed.gz --name Sox10_  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044429)\u001b[0m 2023-01-15 07:53:39,679 cisTopic     INFO     Calling peaks for Macrophages with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Macrophages.bed.gz --name Macrophages  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044426)\u001b[0m 2023-01-15 07:53:39,811 cisTopic     INFO     Calling peaks for Endothelial with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Endothelial.bed.gz --name Endothelial  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044424)\u001b[0m 2023-01-15 07:53:39,810 cisTopic     INFO     Calling peaks for Hepatocyte with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Hepatocyte.bed.gz --name Hepatocyte  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044420)\u001b[0m 2023-01-15 07:53:39,829 cisTopic     INFO     Calling peaks for Adipocytes with /cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2 callpeak --treatment mouse_adrenal/scATAC/consensus_peak_calling/pseudobulk_bed_files/Adipocytes.bed.gz --name Adipocytes  --outdir mouse_adrenal/scATAC/consensus_peak_calling/MACS/ --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044425)\u001b[0m 2023-01-15 07:53:46,810 cisTopic     INFO     Skeletal_muscle done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044424)\u001b[0m 2023-01-15 07:53:48,470 cisTopic     INFO     Hepatocyte done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044431)\u001b[0m 2023-01-15 07:54:02,094 cisTopic     INFO     Smooth_muscle done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044422)\u001b[0m 2023-01-15 07:54:02,862 cisTopic     INFO     Capsule done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044423)\u001b[0m 2023-01-15 07:54:33,188 cisTopic     INFO     Sox10_ done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044429)\u001b[0m 2023-01-15 07:55:16,270 cisTopic     INFO     Macrophages done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044430)\u001b[0m 2023-01-15 07:56:03,457 cisTopic     INFO     Fibroblast done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044428)\u001b[0m 2023-01-15 07:56:38,317 cisTopic     INFO     Stromal done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044420)\u001b[0m 2023-01-15 07:56:49,962 cisTopic     INFO     Adipocytes done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044427)\u001b[0m 2023-01-15 07:59:06,210 cisTopic     INFO     Medulla done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044426)\u001b[0m 2023-01-15 07:59:19,057 cisTopic     INFO     Endothelial done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2044421)\u001b[0m 2023-01-15 08:22:11,520 cisTopic     INFO     Cortex done!\n"
     ]
    }
   ],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import peak_calling\n",
    "\n",
    "# Run peak calling\n",
    "macs_path='/cellar/users/aklie/opt/miniconda3/envs/scenicplus/bin/macs2'\n",
    "narrow_peaks_dict = peak_calling(\n",
    "    macs_path,\n",
    "    bed_paths,\n",
    "    os.path.join(work_dir, 'scATAC/consensus_peak_calling/MACS/'),\n",
    "    genome_size='hs',\n",
    "    n_cpu=12,\n",
    "    input_format='BEDPE',\n",
    "    shift=73,\n",
    "    ext_size=146,\n",
    "    keep_dup = 'all',\n",
    "    q_value = 0.05,\n",
    "    _temp_dir = os.path.join(tmp_dir, 'ray_spill')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dump the return object to a pickle\n",
    "pickle.dump(\n",
    "    narrow_peaks_dict,\n",
    "    open(os.path.join(work_dir, 'scATAC/consensus_peak_calling/MACS/narrow_peaks_dict.pkl'), 'wb')\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to generate a region x cell matrix, we need a consensus peak set. There a lot of ways to do this, again, pycisTopic has a couple functions that we will take advantage of"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pycisTopic.iterative_peak_calling import get_consensus_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-01-15 08:22:15,531 cisTopic     INFO     Extending and merging peaks per class\n",
      "Warning! Start and End columns now have different dtypes: int64 and int32\n",
      "2023-01-15 08:32:59,529 cisTopic     INFO     Done!\n"
     ]
    }
   ],
   "source": [
    "from pycisTopic.iterative_peak_calling import *\n",
    "\n",
    "# Get consensus peaks\n",
    "peak_half_width = 250\n",
    "path_to_blacklist= os.path.join(work_dir, 'hg38-blacklist.v2.bed')\n",
    "consensus_peaks=get_consensus_peaks(\n",
    "    narrow_peaks_dict, \n",
    "    peak_half_width, \n",
    "    chromsizes=chromsizes, \n",
    "    path_to_blacklist=path_to_blacklist\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the consensus regions as a bed file. This will be used for multiple downstream steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "consensus_peaks.to_bed(\n",
    "    path = os.path.join(work_dir, 'scATAC/consensus_peak_calling/consensus_regions.bed'),\n",
    "    keep=True,\n",
    "    compression='infer',\n",
    "    chain=False\n",
    ")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now perform QC on our scATAC data using the consensus peak set. We start by grabbing TSS annotations for protein coding genes. Again remember to choose the correct organism"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pybiomart/dataset.py:269: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  result = pd.read_csv(StringIO(response.text), sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "import pybiomart as pbm\n",
    "\n",
    "# Grab TSS annotations for mouse protein coding genes\n",
    "dataset = pbm.Dataset(name='mmusculus_gene_ensembl',  host='http://www.ensembl.org')\n",
    "annot = dataset.query(attributes=['chromosome_name', 'transcription_start_site', 'strand', 'external_gene_name', 'transcript_biotype'])\n",
    "annot['Chromosome/scaffold name'] = annot['Chromosome/scaffold name'].to_numpy(dtype = str)\n",
    "filter = annot['Chromosome/scaffold name'].str.contains('CHR|GL|JH|MT')\n",
    "annot = annot[~filter]\n",
    "annot['Chromosome/scaffold name'] = annot['Chromosome/scaffold name'].str.replace(r'(\\b\\S)', r'chr\\1')\n",
    "annot.columns=['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "annot = annot[annot.Transcript_type == 'protein_coding']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENCFF187VMN': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed',\n",
       " 'ENCFF035SPT': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed',\n",
       " 'ENCFF622EUO': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed',\n",
       " 'ENCFF119IVK': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed',\n",
       " 'ENCFF683IBE': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed',\n",
       " 'ENCFF042ZJI': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed',\n",
       " 'ENCFF176LJV': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed',\n",
       " 'ENCFF101BLM': 'mouse_adrenal/scATAC/consensus_peak_calling/consensus_regions.bed'}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Each sample will have the same regions here as we want to merge\n",
    "path_to_regions = dict.fromkeys(fragments_dict.keys(), os.path.join(work_dir, 'scATAC/consensus_peak_calling/consensus_regions.bed'))\n",
    "path_to_regions"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next is to calculate a set of metrics for each cell based on consensus regions. Note that this requires counting fragments again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2023-01-15 09:05:14,033 cisTopic     INFO     Reading ENCFF187VMN\n",
      "2023-01-15 09:06:44,590 cisTopic     INFO     Computing barcode rank plot for ENCFF187VMN\n",
      "2023-01-15 09:06:44,591 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:06:50,918 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:06:50,969 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:06:50,973 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:06:57,171 cisTopic     INFO     Computing duplicate rate plot for ENCFF187VMN\n",
      "2023-01-15 09:07:05,086 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:07:05,317 cisTopic     INFO     Computing insert size distribution for ENCFF187VMN\n",
      "2023-01-15 09:07:05,319 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:07:06,860 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:07:56,480 cisTopic     INFO     Computing TSS profile for ENCFF187VMN\n",
      "2023-01-15 09:08:05,050 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:08:05,102 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:08:33,378 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:08:49,005 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:09:00,903 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:09:01,069 cisTopic     INFO     Computing FRIP profile for ENCFF187VMN\n",
      "2023-01-15 09:09:03,023 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:09:14,962 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:10:01,726 cisTopic     INFO     Sample ENCFF187VMN done!\n",
      "2023-01-15 09:10:01,728 cisTopic     INFO     Reading ENCFF035SPT\n",
      "2023-01-15 09:11:28,004 cisTopic     INFO     Computing barcode rank plot for ENCFF035SPT\n",
      "2023-01-15 09:11:28,006 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:11:33,313 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:11:33,357 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:11:33,360 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:11:37,878 cisTopic     INFO     Computing duplicate rate plot for ENCFF035SPT\n",
      "2023-01-15 09:11:45,163 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:11:45,338 cisTopic     INFO     Computing insert size distribution for ENCFF035SPT\n",
      "2023-01-15 09:11:45,340 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:11:46,897 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:12:36,025 cisTopic     INFO     Computing TSS profile for ENCFF035SPT\n",
      "2023-01-15 09:12:45,411 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:12:45,459 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:13:13,605 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:13:24,395 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:13:32,375 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:13:32,542 cisTopic     INFO     Computing FRIP profile for ENCFF035SPT\n",
      "2023-01-15 09:13:34,496 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:13:44,129 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:14:25,542 cisTopic     INFO     Sample ENCFF035SPT done!\n",
      "2023-01-15 09:14:25,543 cisTopic     INFO     Reading ENCFF622EUO\n",
      "2023-01-15 09:16:05,951 cisTopic     INFO     Computing barcode rank plot for ENCFF622EUO\n",
      "2023-01-15 09:16:05,953 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:16:11,914 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:16:11,960 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:16:11,964 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:16:18,361 cisTopic     INFO     Computing duplicate rate plot for ENCFF622EUO\n",
      "2023-01-15 09:16:26,781 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:16:27,011 cisTopic     INFO     Computing insert size distribution for ENCFF622EUO\n",
      "2023-01-15 09:16:27,013 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:16:28,821 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:17:28,724 cisTopic     INFO     Computing TSS profile for ENCFF622EUO\n",
      "2023-01-15 09:17:39,992 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:17:40,042 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:18:13,394 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:18:21,141 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:18:26,911 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:18:27,078 cisTopic     INFO     Computing FRIP profile for ENCFF622EUO\n",
      "2023-01-15 09:18:29,009 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:18:43,220 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:19:36,984 cisTopic     INFO     Sample ENCFF622EUO done!\n",
      "2023-01-15 09:19:36,985 cisTopic     INFO     Reading ENCFF119IVK\n",
      "2023-01-15 09:20:59,989 cisTopic     INFO     Computing barcode rank plot for ENCFF119IVK\n",
      "2023-01-15 09:20:59,991 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:21:04,929 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:21:04,981 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:21:04,983 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:21:09,956 cisTopic     INFO     Computing duplicate rate plot for ENCFF119IVK\n",
      "2023-01-15 09:21:17,068 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:21:17,248 cisTopic     INFO     Computing insert size distribution for ENCFF119IVK\n",
      "2023-01-15 09:21:17,249 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:21:18,703 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:22:03,678 cisTopic     INFO     Computing TSS profile for ENCFF119IVK\n",
      "2023-01-15 09:22:12,131 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:22:12,181 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:22:37,775 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:22:44,568 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:22:49,660 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:22:49,848 cisTopic     INFO     Computing FRIP profile for ENCFF119IVK\n",
      "2023-01-15 09:22:51,839 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:23:01,615 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:23:40,594 cisTopic     INFO     Sample ENCFF119IVK done!\n",
      "2023-01-15 09:23:40,595 cisTopic     INFO     Reading ENCFF683IBE\n",
      "2023-01-15 09:25:19,235 cisTopic     INFO     Computing barcode rank plot for ENCFF683IBE\n",
      "2023-01-15 09:25:19,238 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:25:26,746 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:25:26,801 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:25:26,803 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:25:33,842 cisTopic     INFO     Computing duplicate rate plot for ENCFF683IBE\n",
      "2023-01-15 09:25:43,222 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:25:43,395 cisTopic     INFO     Computing insert size distribution for ENCFF683IBE\n",
      "2023-01-15 09:25:43,396 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:25:45,320 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:26:41,466 cisTopic     INFO     Computing TSS profile for ENCFF683IBE\n",
      "2023-01-15 09:26:51,919 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:26:51,966 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:27:23,943 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:27:33,586 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:27:41,007 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:27:41,171 cisTopic     INFO     Computing FRIP profile for ENCFF683IBE\n",
      "2023-01-15 09:27:43,106 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:27:55,072 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:28:43,158 cisTopic     INFO     Sample ENCFF683IBE done!\n",
      "2023-01-15 09:28:43,159 cisTopic     INFO     Reading ENCFF042ZJI\n",
      "2023-01-15 09:30:16,005 cisTopic     INFO     Computing barcode rank plot for ENCFF042ZJI\n",
      "2023-01-15 09:30:16,007 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:30:21,904 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:30:21,949 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:30:21,952 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:30:27,870 cisTopic     INFO     Computing duplicate rate plot for ENCFF042ZJI\n",
      "2023-01-15 09:30:35,614 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:30:35,819 cisTopic     INFO     Computing insert size distribution for ENCFF042ZJI\n",
      "2023-01-15 09:30:35,820 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:30:37,471 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:31:32,499 cisTopic     INFO     Computing TSS profile for ENCFF042ZJI\n",
      "2023-01-15 09:31:41,788 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:31:41,838 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:32:10,550 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:32:17,043 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:32:22,020 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:32:22,187 cisTopic     INFO     Computing FRIP profile for ENCFF042ZJI\n",
      "2023-01-15 09:32:24,088 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:32:35,391 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:33:19,562 cisTopic     INFO     Sample ENCFF042ZJI done!\n",
      "2023-01-15 09:33:19,563 cisTopic     INFO     Reading ENCFF176LJV\n",
      "2023-01-15 09:35:00,017 cisTopic     INFO     Computing barcode rank plot for ENCFF176LJV\n",
      "2023-01-15 09:35:00,019 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:35:06,992 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:35:07,039 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:35:07,042 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:35:14,203 cisTopic     INFO     Computing duplicate rate plot for ENCFF176LJV\n",
      "2023-01-15 09:35:23,709 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:35:23,921 cisTopic     INFO     Computing insert size distribution for ENCFF176LJV\n",
      "2023-01-15 09:35:23,922 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:35:25,688 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:36:24,368 cisTopic     INFO     Computing TSS profile for ENCFF176LJV\n",
      "2023-01-15 09:36:34,532 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:36:34,581 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:37:06,852 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:37:15,342 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:37:21,946 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:37:22,113 cisTopic     INFO     Computing FRIP profile for ENCFF176LJV\n",
      "2023-01-15 09:37:24,099 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:37:37,612 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:38:26,477 cisTopic     INFO     Sample ENCFF176LJV done!\n",
      "2023-01-15 09:38:26,478 cisTopic     INFO     Reading ENCFF101BLM\n",
      "2023-01-15 09:39:56,674 cisTopic     INFO     Computing barcode rank plot for ENCFF101BLM\n",
      "2023-01-15 09:39:56,676 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:40:02,766 cisTopic     INFO     Marking barcodes with more than 100\n",
      "2023-01-15 09:40:02,819 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:40:02,821 cisTopic     INFO     Returning valid barcodes\n",
      "2023-01-15 09:40:08,791 cisTopic     INFO     Computing duplicate rate plot for ENCFF101BLM\n",
      "2023-01-15 09:40:16,714 cisTopic     INFO     Return plot data\n",
      "2023-01-15 09:40:16,898 cisTopic     INFO     Computing insert size distribution for ENCFF101BLM\n",
      "2023-01-15 09:40:16,899 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:40:18,473 cisTopic     INFO     Returning plot data\n",
      "2023-01-15 09:41:10,246 cisTopic     INFO     Computing TSS profile for ENCFF101BLM\n",
      "2023-01-15 09:41:19,734 cisTopic     INFO     Formatting annnotation\n",
      "2023-01-15 09:41:19,782 cisTopic     INFO     Creating coverage matrix\n",
      "2023-01-15 09:41:48,990 cisTopic     INFO     Coverage matrix done\n",
      "2023-01-15 09:41:57,548 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "2023-01-15 09:42:04,150 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "2023-01-15 09:42:04,316 cisTopic     INFO     Computing FRIP profile for ENCFF101BLM\n",
      "2023-01-15 09:42:06,259 cisTopic     INFO     Counting fragments\n",
      "2023-01-15 09:42:17,425 cisTopic     INFO     Intersecting fragments with regions\n",
      "2023-01-15 09:43:01,085 cisTopic     INFO     Sample ENCFF101BLM done!\n"
     ]
    }
   ],
   "source": [
    "from pycisTopic.qc import *\n",
    "\n",
    "# Calculate QC stats for each sample\n",
    "metadata_bc, profile_data_dict = compute_qc_stats(\n",
    "    fragments_dict = fragments_dict,\n",
    "    tss_annotation = annot,\n",
    "    stats=['barcode_rank_plot', 'duplicate_rate', 'insert_size_distribution', 'profile_tss', 'frip'],\n",
    "    label_list = None,\n",
    "    path_to_regions = path_to_regions,\n",
    "    n_cpu = 1,\n",
    "    valid_bc = None,\n",
    "    n_frag = 100,\n",
    "    n_bc = None,\n",
    "    tss_flank_window = 1000,\n",
    "    tss_window = 50,\n",
    "    tss_minimum_signal_window = 100,\n",
    "    tss_rolling_window = 10,\n",
    "    remove_duplicates = True,\n",
    "    _temp_dir = os.path.join(tmp_dir + 'ray_spill')\n",
    ")\n",
    "\n",
    "# Dump the qc results in some objects\n",
    "if not os.path.exists(os.path.join(work_dir, 'scATAC/quality_control')):\n",
    "    os.makedirs(os.path.join(work_dir, 'scATAC/quality_control'))\n",
    "\n",
    "pickle.dump(metadata_bc,\n",
    "            open(os.path.join(work_dir, 'scATAC/quality_control/metadata_bc.pkl'), 'wb'))\n",
    "\n",
    "pickle.dump(profile_data_dict,\n",
    "            open(os.path.join(work_dir, 'scATAC/quality_control/profile_data_dict.pkl'), 'wb'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can plot a few QC plots and see how filtering would affect the cells called"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF187VMN\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF035SPT\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF622EUO\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF119IVK\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF683IBE\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF042ZJI\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF176LJV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample ENCFF101BLM\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1871: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n",
      "/cellar/users/aklie/opt/miniconda3/envs/scenicplus/lib/python3.8/site-packages/pycisTopic/qc.py:1886: UserWarning: \n",
      "\n",
      "`distplot` is a deprecated function and will be removed in seaborn v0.14.0.\n",
      "\n",
      "Please adapt your code to use either `displot` (a figure-level function with\n",
      "similar flexibility) or `histplot` (an axes-level function for histograms).\n",
      "\n",
      "For a guide to updating your code to use the new functions, please see\n",
      "https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751\n",
      "\n",
      "  sns.distplot(\n"
     ]
    }
   ],
   "source": [
    "from pycisTopic.qc import *\n",
    "\n",
    "# Perform some filtering to keep only high quality cells\n",
    "QC_filters = {\n",
    "    'Log_unique_nr_frag': [2 , None],\n",
    "    'FRIP':               [0.3, None],\n",
    "    'TSS_enrichment':     [1   , None],\n",
    "    'Dupl_rate':          [None, None]\n",
    "\n",
    "}\n",
    "bc_passing_filters = {}\n",
    "for sample in fragments_dict:\n",
    "    print(\"Sample\", sample)\n",
    "    FRIP_NR_FRAG_filter = plot_barcode_metrics(\n",
    "        metadata_bc[sample],\n",
    "        var_x='Log_unique_nr_frag',\n",
    "        var_y='FRIP',\n",
    "        min_x=QC_filters['Log_unique_nr_frag'][0],\n",
    "        max_x=QC_filters['Log_unique_nr_frag'][1],\n",
    "        min_y=QC_filters['FRIP'][0],\n",
    "        max_y=QC_filters['FRIP'][1],\n",
    "        return_cells=True,\n",
    "        return_fig=False,\n",
    "        plot=False\n",
    "    )\n",
    "    # Return figure to plot together with other metrics, and cells passing filters\n",
    "    TSS_NR_FRAG_filter = plot_barcode_metrics(\n",
    "        metadata_bc[sample],\n",
    "        var_x='Log_unique_nr_frag',\n",
    "        var_y='TSS_enrichment',\n",
    "        min_x=QC_filters['Log_unique_nr_frag'][0],\n",
    "        max_x=QC_filters['Log_unique_nr_frag'][1],\n",
    "        min_y=QC_filters['TSS_enrichment'][0],\n",
    "        max_y=QC_filters['TSS_enrichment'][1],\n",
    "        return_cells=True,\n",
    "        return_fig=False,\n",
    "        plot=False\n",
    "    )\n",
    "    bc_passing_filters[sample] = list((set(FRIP_NR_FRAG_filter) & set(TSS_NR_FRAG_filter)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3887 barcodes passed QC stats for ENCFF187VMN\n",
      "5623 barcodes passed QC stats for ENCFF035SPT\n",
      "5210 barcodes passed QC stats for ENCFF622EUO\n",
      "4669 barcodes passed QC stats for ENCFF119IVK\n",
      "6547 barcodes passed QC stats for ENCFF683IBE\n",
      "4875 barcodes passed QC stats for ENCFF042ZJI\n",
      "6208 barcodes passed QC stats for ENCFF176LJV\n",
      "5236 barcodes passed QC stats for ENCFF101BLM\n"
     ]
    }
   ],
   "source": [
    "# Dump and print the cells that passed!\n",
    "pickle.dump(\n",
    "    bc_passing_filters,\n",
    "    open(os.path.join(work_dir, 'scATAC/quality_control/bc_passing_filters.pkl'), 'wb')\n",
    ")\n",
    "for sample in bc_passing_filters:\n",
    "    print(f\"{len(bc_passing_filters[sample])} barcodes passed QC stats for {sample}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8 scenicplus",
   "language": "python",
   "name": "scenicplus"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
